### PR TITLE
Automate validation of pipeline configs

### LINKED ISSUE #131: Add a schema for validating pipeline configuration
When we load a taxonomy yaml file, we validate its contents against a schema contained in `instructlab.schema`.  Here is an example schema:

https://github.com/instructlab/schema/blob/main/src/instructlab/schema/v2/knowledge.json

There's some code for loading this schema and validating with it in `instructlab.sdg.utils.taxonomy`.

It would be nice if we did something similar when we load a pipeline yaml. It can help catch subtle mistakes. We could also provide instructions for how to check a configuration against the schema manually before trying to run it.

Here is a start at what a schema could look like for pipeline configs (auto-generated, not tested yet).

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "version": {
      "type": "string"
    },
    "blocks": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          },
          "type": {
            "type": "string"
          },
          "config": {
            "type": "object",
            "properties": {
              "config_path": {
                "type": "string"
              },
              "output_cols": {
                "type": "array",
                "items": {
                  "type": "string"
                }
              }
            },
            "required": ["config_path", "output_cols"]
          },
          "gen_kwargs": {
            "type": "object",
            "properties": {
              "temperature": {
                "type": "number"
              },
              "max_tokens": {
                "type": "integer"
              },
              "n": {
                "type": "integer"
              }
            }
          },
          "drop_duplicates": {
            "type": "array",
            "items": {
              "type": "string"
            }
          }
        },
        "required": ["name", "type", "config"]
      }
    }
  },
  "required": ["version", "blocks"]
}
```

### PR DESCRIPTION
**Pull request by russellb**
Closes #131 

1cfd4bb Add jsonschema for validating pipeline configuration
c3bd917 Add a script that will validate pipeline configurations
18b4741 Add tox env for validating pipeline configuration
3090e65 Run pipeline config validation in CI

commit 1cfd4bbd494ca936ce7e072f498dc208b8e0dfec
Author: Russell Bryant <rbryant@redhat.com>
Date:   Sun Jul 14 12:50:45 2024 -0400

    Add jsonschema for validating pipeline configuration
    
    This commit introduces a `v1.json` jsonschema for use with validating
    pipeline configuration. This follows the same format used in the
    `instructlab/schema` repository which we use for validating taxonomy
    `qna.yaml` files.
    
    It's very likely that the schema is missing some details that are
    allowed. It does pass against all configurations currently in the
    source tree, though.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit c3bd917e55dd66f241da152d782e783385742294
Author: Russell Bryant <rbryant@redhat.com>
Date:   Sun Jul 14 12:53:43 2024 -0400

    Add a script that will validate pipeline configurations
    
    This script loads the pipeline validation schema and validates all
    configurations under `src/instructlab/sdg/pipelines/`.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 18b47418651b8d1f3f5c000a5477894398662501
Author: Russell Bryant <rbryant@redhat.com>
Date:   Sun Jul 14 12:52:42 2024 -0400

    Add tox env for validating pipeline configuration
    
    To run the script that validates pipeline configurations, you may now
    run:
    
        tox -e validate-pipelines
    
    or
    
        make validate-pipelines
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 3090e657515a0782260644cd63ba9513fa32b920
Author: Russell Bryant <rbryant@redhat.com>
Date:   Sun Jul 14 13:00:03 2024 -0400

    Run pipeline config validation in CI
    
    Updating the existing linting job to include pipeline config
    validation, as well.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

### CODE REVIEW

**Comment by russellb** [comment]
Also, a follow-up enhancement to this would be to utilize the schema at runtime to validate custom schemas as soon as they are loaded and prior to trying to execute them.

File: scripts/validate_pipelines.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        validate(instance=pipeline, schema=schema)
+        print(f"Validation successful for {yaml_file}.")
+    except jsonschema.exceptions.ValidationError as err:
+        print(f"Validation failed for {yaml_file}:", err)
+        return False
+    return True
+
+
+def main():
+    schema_path = "src/instructlab/sdg/pipelines/schema/v1.json"
```
Comment (markmc) [inline scripts/validate_pipelines.py:28]: How about loading this from the `instructlab.sdg` package

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          },
+          "drop_columns": {
+            "type": "array",
+            "items": {
+              "type": "string"
+            }
+          },
+          "gen_kwargs": {
+            "type": "object",
+            "properties": {
```
Comment (markmc) [inline src/instructlab/sdg/pipelines/schema/v1.json:37]: `model` is also supported here - this is how you can point to a non-default model

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            }
+          },
+          "config": {
+            "anyOf": [
+              {
+                "type": "object",
+                "description": "FilterByValueBlock",
+                "additionalProperties": true,
+                "properties": {
+                  "config_path": {
```
Comment (markmc) [inline src/instructlab/sdg/pipelines/schema/v1.json:56]: LLMBlock only

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                    "type": "string"
+                  },
+                  "output_cols": {
```
Comment (markmc) [inline src/instructlab/sdg/pipelines/schema/v1.json:59]: LLMBlock only

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                    "type": "array",
+                    "items": {
+                      "type": "string"
+                    }
+                  },
+                  "convert_dtype": {
+                    "type": "string"
```
Comment (markmc) [inline src/instructlab/sdg/pipelines/schema/v1.json:66]: We only support float, int, and bool here

Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json:66]: `"enum": ["float", "int", "bool"]`

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                      {
+                        "type": "string"
+                      },
+                      {
+                        "type": "number"
+                      }
+                    ]
+                  },
+                  "operation": {
+                    "type": "string"
```
Comment (markmc) [inline src/instructlab/sdg/pipelines/schema/v1.json:82]: You could list the supported ones?

Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json:82]: `"enum": ["eq", "ne", "gt", "ge", "lt", "le", "contains"]`

but I think you're trying to get rid of `contains`?

File: src/instructlab/sdg/pipelines/schema/v1.json (hunk 2 above)
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json:37]: Hmm, this should really be an LLMBlock thing - it's only used there, and it's really arguments for the OpenAI `completions.create()` call

Filed as #138

Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json:37]: I've added model now

File: src/instructlab/sdg/pipelines/schema/v1.json (hunk 3 above)
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json]: Fixed

File: src/instructlab/sdg/pipelines/schema/v1.json (hunk 4 above)
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json]: Fixed

File: src/instructlab/sdg/pipelines/schema/v1.json (hunk 5 above)
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json]: Fixed

File: src/instructlab/sdg/pipelines/schema/v1.json (hunk 6 above)
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json]: Fixed

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+              "n": {
+                "type": "number"
+              }
+            }
+          },
+          "config": {
+            "anyOf": [
+              {
+                "type": "object",
+                "description": "FilterByValueBlock",
```
[the remark this answers is not in the document]
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json:59]: You have it down below `"required": ["filter_column", "filter_value", "operation"]`

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                "additionalProperties": true,
+                "properties": {
+                  "config_path": {
+                    "type": "string"
+                  },
```
[the remark this answers is not in the document]
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json]: Right, fixed

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                  "operation": {
+                    "type": "string"
+                  }
+                },
+                "required": ["filter_column", "filter_value", "operation"]
+              },
+              {
+                "type": "object",
+                "description": "LLMBlock",
```
[the remark this answers is not in the document]
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json:112]: Fixed

File: src/instructlab/sdg/pipelines/schema/v1.json [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                        "type": "string"
+                      },
+                      {
+                        "type": "number"
+                      }
+                    ]
+                  },
+                  "operation": {
+                    "type": "string"
+                  }
```
[the remark this answers is not in the document]
Comment (markmc) [reply src/instructlab/sdg/pipelines/schema/v1.json]: Added

**Review by markmc — APPROVED** [review]
My main outstanding comment is about making this a bit more general purpose, but that can come later

I think this is good to go ... I'm fine if you'd prefer to squash all my commits though

File: scripts/validate_pipelines.py (hunk 1 above)
Comment (russellb) [reply scripts/validate_pipelines.py:28]: yes, that would be better

File: scripts/validate_pipelines.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    with open(schema_path, "r") as file:
+        schema = json.load(file)
+
+    yaml_files = glob.glob("src/instructlab/sdg/pipelines/**/*.yaml", recursive=True)
```
[the remark this answers is not in the document]
Comment (russellb) [reply scripts/validate_pipelines.py:32]: Yeah, that sounds good.

We could even leave this as the default when you don't pass an argument

**Comment by russellb** [comment]
> My main outstanding comment is about making this a bit more general purpose, but that can come later
> 
> I think this is good to go ... I'm fine if you'd prefer to squash all my commits though

Thanks for the additions!  lgtm.

I'm going to merge as having reviewed each others additions to the PR.

**Comment by russellb** [comment]
filed this for making the validation script more generally useful -- https://github.com/instructlab/sdg/issues/139

**Pull request merged**